### PR TITLE
Use `@plutojl/lang-julia` for CodeMirror 6 Julia support

### LINKED ISSUE #13943: Highlight function calls for Julia
### Problem

I have a problem with syntax highlight for Julia: function calls are not highlighted in any way. This is strange as they are highlighted in Jupyter notebook so I hope it is not a hard fix. The issue is described here: https://discourse.julialang.org/t/modify-syntax-highlighting-for-ijulia-in-jupyter-lab-to-resemble-jupyter-notebook/82168. 

### Proposed Solution

Highlight function calls in light green like in Jupyter notebooks. 

### Additional context

I have Jupyter Lab 3.6.1 and IJulia 1.24, which I believe are the latest versions.

### LINKED ISSUE #19046: New CodeMirror 6 language mode for Julia
Hi! 👋 I'm Fons, I work on Julia dev tooling, and the [Pluto.jl](https://plutojl.org/) Julia notebook environment. I presented Pluto at [JupyterCon 2023](https://www.youtube.com/watch?v=Rg3r3gG4nQo).

The Pluto team has developed a new lezer parser for Julia, which adds direct support for CodeMirror 6. [`@plutojl/lezer-julia`](https://github.com/JuliaPluto/lezer-julia) and [`@plutojl/lang-julia`](https://github.com/JuliaPluto/lang-julia). We have been using it in production for some years already in Pluto.jl, and I think it's in a good enough state to also use in JupyterLab! I recently released version `1.0.0` of `@plutojl/lang-julia` to make this official. We would feel very proud to contribute to JupyterLab, and we think this would be a valuable contribution to making Julia more fun and accessible to use.

# Advantages
The new parser is much more detailed than the old one. This is clearly visible in syntax highlighting, where you just see a lot more useful colors for different Julia features. Even on Julia basics there is a big difference - it adds support for functions, lists, type annotations, symbols, string interpolation and `.property` access. And it's more correct: it has been tested on countless quirky edge cases with good results.

It also adds support for all new Julia features (currently up and until Julia 1.13), while the legacy parser has not been updated for years. Our parser is actively maintained.

See for yourself! I set up this demo site (vibe coded) where you can see the same code with the `@plutojl/lang-julia` parser and the legacy mode:  https://juliapluto.github.io/lang-julia-demo/

Screenshot of the two parsers in action. The new parser shows a lot more variation in color and style, while the old parser looks fairly bland.

_The new parser (left) vs the legacy mode parser (right)._

# Disadvantages
The bundle size is 297kB, which would add to the jupyterlab bundle size. It's quite large because Julia has very complex syntax, which makes the Julia lezer one of the more complex parsers in the CM ecosystem. 

I have not compared the difference in editing performance. The new parser might be faster, because it's better integrated into CM6. But it might also be slower because of its higher complexity. Either case, it's a high-performance parser, even on low-end devices, and CM6's incremental parsing works really well with it.

# Suggestion
I think that `@plutojl/lang-julia` would be a cool addition to JupyterLab! It should be plug-and-play, and you get a lot of new syntax highlighting automatically without any changes. In Pluto (and the demo website), we added more context-based syntax highlighting (which is now possible, since you get a full parse tree) like `t.function(t.definition(t.variableName))` for coloring the name of a function definition. You could add this to JupyterLab too, but I imagine you might want to keep the syntax highlighting code language-independent.

What do you think? I would be happy (and excited) to work on a PR.

# AI disclaimer
The `@plutojl/lezer-julia` parser was written by hand over the past 5 years, with a couple AI-assisted (human reviewed) fixes in recent years. 
The demo website is vibe coded. 
This post is hand written.

### PR DESCRIPTION
**Pull request by fonsp**
This implements the new CodeMirror 6 parser that I proposed in #19046. 

## References

Fix #19046 

## Code changes

Add the `@plutojl/lang-julia` dependency and wire it in.

## User-facing changes

Before:
Scherm­afbeelding 2026-06-18 om 16 55 22

After:
Scherm­afbeelding 2026-06-19 om 12 46 00

## Backwards-incompatible changes

No

## AI usage

- NO: Some or all of the content of this PR was generated by AI.
- YES: The human author has carefully reviewed this PR and run this code (keep this PR "draft" until the answer is YES)

### CODE REVIEW

**Comment by fonsp** [comment]
<details>
<summary>Current state of this PR:</summary>


Scherm­afbeelding 2026-06-18 om 16 51 55

</details>




I think it could be a lot better - there is much more information available in the syntax tree, but they are currently not used for coloring. See my screenshot in https://github.com/jupyterlab/jupyterlab/issues/19046 for an example. I will look into this next week!

**Comment by fonsp** [comment]
Okay, I made some improvements:
1. In `@plutojl/lezer-julia`, I declared more style tags ([PR](https://github.com/JuliaPluto/lezer-julia/pull/59)). Useful for this PR: I added `function`, which will highlight function calls and definitions (just like with the legacy grammar, or with other languages).
2. In JupyterLab, I added highlighting for two CM tags, which were previously not being highlighted: `macroName` and `typeName`. I reused existing colors for these, to keep this PR simple.


Note: The change 2. makes a big difference for Julia, which uses a lot of types and macros. The change will also affect other languages with these features, the first one that came to mind is Rust (typed language). I think this side effect is positive: languages with types and macros will now have more detailed/diverse highlighting.

This PR is now ready for review!


<details>
<summary>Current state of this PR (updated the PR description with this image).</summary>

Scherm­afbeelding 2026-06-19 om 12 46 00


</details>

File: packages/codemirror/src/theme.ts [hunk 1]
```diff
@@ -138,15 +138,15 @@ export const jupyterHighlightStyle = HighlightStyle.define([
   { tag: t.atom, color: 'var(--jp-mirror-editor-atom-color)' },
   { tag: t.number, color: 'var(--jp-mirror-editor-number-color)' },
   {
-    tag: [t.definition(t.name), t.function(t.definition(t.variableName))],
+    tag: t.definition(t.name),
```
Comment (fonsp) [inline packages/codemirror/src/theme.ts:141]: This is just a cleanup: the pattern `t.function(t.definition(t.variableName))` is a strict subset of `t.definition(t.name)`, so I removed it. This is because `t.variableName` is a special case of `t.name`.

File: packages/codemirror/src/theme.ts [hunk 2]
```diff
@@ -138,15 +138,15 @@ export const jupyterHighlightStyle = HighlightStyle.define([
...
     color: 'var(--jp-mirror-editor-def-color)'
   },
   {
-    tag: t.standard(t.variableName),
+    tag: [t.standard(t.variableName), t.typeName],
```
Comment (fonsp) [inline packages/codemirror/src/theme.ts:145]: This adds highlighting for `typeName`, which was previously not highlighted (shown in black). This has side effects on other typed programming languages like Rust, but I think it's a positive change for those languages too.

File: packages/codemirror/src/theme.ts [hunk 3]
```diff
@@ -138,15 +138,15 @@ export const jupyterHighlightStyle = HighlightStyle.define([
...
     color: 'var(--jp-mirror-editor-builtin-color)'
   },
   {
-    tag: [t.special(t.variableName), t.self],
+    tag: [t.special(t.variableName), t.self, t.macroName],
```
Comment (fonsp) [inline packages/codemirror/src/theme.ts:154]: Same comment as for `typeName` above.

**Comment by krassowski** [comment]
Testing this I see:

| Before | After |
|--|--|
| image | image

It looks like an improvement overall, but one think I notice in line 16, 17 and 20 is lost highlight (`struct Circle`, `type Shape` and `mutable struct Point`) - is this intended?

**Comment by fonsp** [comment]
Hi! I took some time for other projects, but I'm back :)

The missing type highlighting was on purpose, limitation in our parser. But I now [added](https://github.com/JuliaPluto/lezer-julia/pull/60/changes) a special node type to our parser to make it possible to highlight type definitions, so this is now possible! I updated this PR to use the new version of `lang-julia`.

<details>
<summary>Screenshot with lang-julia 1.2.0</summary>

Scherm­afbeelding 2026-07-02 om 15 43 25 1


</details>

I also made an extra commit c6de2252b57486ba0bde8f4e32cc441cae75be51 to add special highlighting for the type definition (bold, definition color). Then we get this:

<details>
<summary>Screenshot with lang-julia 1.2.0 and added highlighting rule</summary>

Scherm­afbeelding 2026-07-02 om 15 43 50


</details>

I like it 😄 since type definitions are very important in Julia code, so they deserve to stand out. But feel free to revert that commit for whatever reason. (Same goes for my previous commit cb95db55d86bf189bb428dc67cd9175bbc4c10b4 adding new styles.)

Let me know what you think! :))

**Comment by krassowski** [comment]
Thanks, this looks well! Just a short lint pass and test update to make CI pass and I think we can merge

**Comment by krassowski** [comment]
Final state:

| Before | After |
|--|--|
| image | image

**Review by krassowski — APPROVED** [review]
Thank you!
